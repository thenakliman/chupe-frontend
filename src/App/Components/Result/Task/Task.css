
.task-result-container {
  margin: 2% 20%;
}

.task-create-container {
  display: flex;
  justify-content: space-between
}

.task-create-button {
  padding: 0.75%;
  color: #fff;
  border: 1px solid #28a745;
  background-color: #28a745;
  background-image: linear-gradient(-180deg, #34d058 0%, #28a745);
  cursor:pointer;
}

.task-input-description {
  width: 90%;
  padding: 0.5rem;
}

.team-fund-table-id {
  margin: 4% 0%;
}