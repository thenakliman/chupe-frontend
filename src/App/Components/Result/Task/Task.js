import propTypes from 'prop-types';
import React from 'react';
import * as constants from '../constants';
require('./Task.css');

/**
* Task result component of the application.
*
* @author: thenakliman
*/
export class Task extends React.Component {
  /** Constructor for the component.
   *  @param {object} props
   */
  constructor(props) {
    super(props);
    this.handleDescriptionChange = this.handleDescriptionChange.bind(this);
    this.handleChangeInState = this.handleChangeInState.bind(this);
    this.getInitialState = this.getInitialState.bind(this);
    this.createTask = this.createTask.bind(this);

    this.state = this.getInitialState();
  }

  /** Update state for the change in description of the task.
   * @param {string} value of the description
   */
  handleDescriptionChange(value) {
    this.setState({description: value});
  }

  /** Fetch tasks from the backend and update in the store. */
  componentDidMount() {
    this.props.getTasks();
  }

  /** Get initial state.
   * @return {object} description of the question
   */
  getInitialState() {
    return {description: ''};
  }

  /** Handle change in state.
   * @param {string} id of the question
   * @param {String} newState to be updated
   */
  handleChangeInState(id, newState) {
    const task = this.props.tasks.find((task) => task.id === id);
    this.props.updateTask({...task, state: newState});
  }

  /** Create task on click of create button. */
  createTask() {
    if (this.state.description === '' || this.state.description === undefined) {
      return;
    }

    const task = {
      description: this.state.description,
      progress: 0,
      state: constants.CREATED,
      createdBy: this.props.currentUser,
    };

    this.props.createTask(task);
    this.setState({description: ''});
  }
/**
* Task result component of the application.
*
* @return {Object} TaskResult component.
*/
  render() {
    const states = [
        constants.CREATED,
        constants.IN_PROGRESS,
        constants.ON_HOLD,
        constants.DONE,
    ];

    const sortedTasks = [...this.props.tasks];
    sortedTasks.sort((task1, task2) => task1.id<task2.id?-1:1);

    return (
        <div className='task-result-container'>
          <span id='task-create-container' className={'task-create-container'}>
            <input id='task-input-description'
                   className='task-input-description'
                   value={this.state.description}
                   onChange={
                      (e) => this.handleDescriptionChange(e.target.value)}/>

            <button id='task-create-button'
                    className='task-create-button'
                    onClick={this.createTask}>
              Create
            </button>
          </span>
          <table id='team-fund-table-id'
                 className='team-fund-table-id'>
            <thead>
              <tr>
                <th>S.No</th>
                <th>Description</th>
                <th>Created</th>
                <th>In Progress</th>
                <th>On Hold</th>
                <th>Done</th>
              </tr>
            </thead>
            <tbody>
              {
                sortedTasks.map((task, index) => (
                  <tr key={index}>
                    <td>{index}</td>
                    <td>{task.description}</td>
                    {
                      states.map((state) => (
                        <td key={state}>
                          <input
                            className={'task-state-radio-button'}
                            id={`task-state-radio-button-${state}-${task.id}`}
                            type="radio"
                            name={`state-${index}`}
                            onChange={
                              () => this.handleChangeInState(task.id, state)}
                            checked={state === task.state} />
                        </td>
                      ))
                  }
                  </tr>
                ))
              }
            </tbody>
          </table>
        </div>
    );
  }
}

Task.propTypes = {
  tasks: propTypes.arrayOf(propTypes.object).isRequired,
  currentUser: propTypes.string.isRequired,
  getTasks: propTypes.func.isRequired,
  createTask: propTypes.func.isRequired,
  updateTask: propTypes.func.isRequired,
};
