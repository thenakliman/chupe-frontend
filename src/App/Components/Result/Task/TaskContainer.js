import {Task} from './Task';
import {connect} from 'react-redux';
import {getAllTasks} from '../../../Actions/taskActions';

export const mapStateToProps = (state) => ({
    tasks: state.tasks,
});

export const mapDispatchToProps = (dispatch) => ({
    getTasks: () => {
        dispatch(getAllTasks());
    },
});

export default connect(mapStateToProps, mapDispatchToProps)(Task);
