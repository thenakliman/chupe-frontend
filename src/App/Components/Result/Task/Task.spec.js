import React from 'react'; // eslint-disable-line no-unused-vars
import {Task} from './Task'; // eslint-disable-line no-unused-vars
import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';

describe('Task result component snapshot', () => {
    it('should match the snapshot', () => {
        const tasks = [{
            description: 'task1',
            state: 'IN_PROGRESS',
          }, {
            description: 'task2',
            state: 'IN_PROGRESS',
          }];
        const wrapper = shallow(<Task tasks={tasks}
                                      currentUser={'lal_singh'}
                                      updateTask={jest.fn()}
                                      createTask={() => {}}
                                      getTasks={()=>{}}/>);

        expect(toJson(wrapper)).toMatchSnapshot();
    });
});

describe('Task result component', () => {
    it('Display list of tasks', () => {
        const tasks = [
            {description: 'task1', state: 'IN_PROGRESS'},
            {description: 'task2', state: 'IN_PROGRESS'},
            {description: 'task3', state: 'IN_PROGRESS'},
        ];
        const wrapper = shallow(
            <Task tasks={tasks}
                  updateTask={jest.fn()}
                  getTasks={()=>{}}
                  currentUser={'lal_singh'}
                  createTask={() => {}}/>);

        expect(wrapper.find('tbody').children().length).toEqual(3);
    });

    it('Component did mount', () => {
        const tasks = [{'taskname': 'task1', 'state': 'IN_PROGRESS'}];
        const getTasks = jest.fn();
        shallow(<Task tasks={tasks}
                      getTasks={getTasks}
                      updateTask={jest.fn()}
                      currentUser={'lal_singh'}
                      createTask={() => {}}/>);

        expect(getTasks).toHaveBeenCalledWith();
    });

    it('Should have task description input field', () => {
        const tasks = [{'taskname': 'task1', 'state': 'IN_PROGRESS'}];
        const getTasks = jest.fn();
        const wrapper = shallow(<Task tasks={tasks}
                                      getTasks={getTasks}
                                      updateTask={jest.fn()}
                                      currentUser={'lal_singh'}
                                      createTask={() => {}}/>);

        expect(wrapper.find('#task-input-description').length).toBe(1);
    });

    it('Should have task description input field', () => {
        const tasks = [{'taskname': 'task1', 'state': 'IN_PROGRESS'}];
        const getTasks = jest.fn();
        const wrapper = shallow(<Task tasks={tasks}
                                      updateTask={jest.fn()}
                                      getTasks={getTasks}
                                      currentUser={'lal_singh'}
                                      createTask={() => {}}/>);
        expect(wrapper.find('#task-create-button').length).toBe(1);
    });

    it('should store input value change in component state', () => {
        const tasks = [{'taskname': 'task1', 'state': 'IN_PROGRESS'}];
        const getTasks = jest.fn();
        const wrapper = shallow(<Task tasks={tasks}
                                      getTasks={getTasks}
                                      updateTask={jest.fn()}
                                      currentUser={'lal_singh'}
                                      createTask={() => {}}/>);
        const taskDescription = 'task description';
        wrapper.find('#task-input-description').simulate('change',
            {target: {value: taskDescription}});

        expect(wrapper.state()).toEqual({
            description: taskDescription,
            displayAllTask: false});
    });

    it('should store input value change in component state', () => {
        const tasks = [{'taskname': 'task1', 'state': 'IN_PROGRESS'}];
        const getTasks = jest.fn();
        const wrapper = shallow(
            <Task tasks={tasks}
                  updateTask={jest.fn()}
                  getTasks={getTasks}
                  currentUser={'lal_singh'}
                  createTask={() => {}}/>);

        const taskDescription = 'task description';
        wrapper.find('#task-input-description').simulate('change',
            {target: {value: taskDescription}});

        expect(wrapper.state()).toEqual({
          description: taskDescription,
          displayAllTask: false,
        });
    });

    it('should call create task on click of create button', () => {
        const tasks = [{'taskname': 'task1'}];
        const getTasks = jest.fn();
        const createTask = jest.fn();
        const username = 'nakli';
        const wrapper = shallow(
            <Task tasks={tasks}
                  getTasks={getTasks}
                  updateTask={jest.fn()}
                  createTask={createTask}
                  currentUser={username}/>);

        const taskDescription = 'task description';
        wrapper.find('#task-input-description').simulate('change',
            {target: {value: taskDescription}});

        wrapper.find('#task-create-button').simulate('click');
        expect(createTask).toHaveBeenCalledWith({
            createdBy: username,
            description: taskDescription,
            progress: 0,
            state: 'CREATED',
        });
        expect(wrapper.state()).toEqual({
            description: '',
            displayAllTask: false,
        });
    });

    it('should call create task on click of create button', () => {
        const tasks = [{'taskname': 'task1'}];
        const createTask = jest.fn();
        const wrapper = shallow(<Task tasks={tasks}
                                      getTasks={jest.fn()}
                                      updateTask={jest.fn()}
                                      createTask={createTask}
                                      currentUser={'fake'}
                                />);

        wrapper.find('#task-create-button').simulate('click');
        expect(createTask).not.toHaveBeenCalled();
    });

    it('should call update task on click of CREATE button', () => {
        const tasks = [
            {description: 'task1', id: 2, state: 'IN_PROGRESS'},
            {description: 'task1', id: 1, state: 'IN_PROGRESS'},
            {description: 'task1', id: 3, state: 'IN_PROGRESS'},
        ];

        const updateTask = jest.fn();
        const wrapper = shallow(<Task tasks={tasks}
                                      getTasks={jest.fn()}
                                      updateTask={updateTask}
                                      createTask={jest.fn()}
                                      currentUser={'fake'}
                                />);

        wrapper.find('#task-state-radio-button-CREATED-1').simulate('change');
        expect(updateTask).toHaveBeenCalledWith({
            description: 'task1',
            id: 1,
            state: 'CREATED'});
    });

    it('should show only DONE and IN_PROGRESS tasks', () => {
        const tasks = [
            {description: 'task1', id: 2, state: 'CREATED'},
            {description: 'task1', id: 1, state: 'DONE'},
            {description: 'task1', id: 3, state: 'IN_PROGRESS'},
            {description: 'task1', id: 4, state: 'ON_HOLD'},
        ];

        const updateTask = jest.fn();
        const wrapper = shallow(<Task tasks={tasks}
                                      getTasks={jest.fn()}
                                      updateTask={updateTask}
                                      createTask={jest.fn()}
                                      currentUser={'fake'}
                                />);

        /** two from task and 1 from header row */
        expect(wrapper.find('tr').length).toBe(3);
    });

    it('should have button to display all tasks', () => {
        const tasks = [
            {description: 'task1', id: 2, state: 'CREATED'},
            {description: 'task1', id: 1, state: 'CREATED'},
            {description: 'task1', id: 3, state: 'CREATED'},
            {description: 'task1', id: 4, state: 'CREATED'},
        ];

        const updateTask = jest.fn();
        const wrapper = shallow(<Task tasks={tasks}
                                      getTasks={jest.fn()}
                                      updateTask={updateTask}
                                      createTask={jest.fn()}
                                      currentUser={'fake'}
                                />);

        expect(wrapper.find('#tasks-button').length).toBe(1);
    });

    it('should display all task on click of all task button', () => {
        const tasks = [
            {description: 'task1', id: 2, state: 'CREATED'},
            {description: 'task1', id: 1, state: 'DONE'},
            {description: 'task1', id: 3, state: 'IN_PROGRESS'},
            {description: 'task1', id: 4, state: 'ON_HOLD'},
        ];

        const updateTask = jest.fn();
        const wrapper = shallow(<Task tasks={tasks}
                                      getTasks={jest.fn()}
                                      updateTask={updateTask}
                                      createTask={jest.fn()}
                                      currentUser={'fake'}
                                />);

        wrapper.find('#tasks-button').simulate('click');
        expect(wrapper.find('tr').length).toBe(5);
    });

    it('should display pending task on click of pending tasks button', () => {
        const tasks = [
            {description: 'task1', id: 2, state: 'CREATED'},
            {description: 'task1', id: 1, state: 'DONE'},
            {description: 'task1', id: 3, state: 'IN_PROGRESS'},
            {description: 'task1', id: 4, state: 'ON_HOLD'},
        ];

        const updateTask = jest.fn();
        const wrapper = shallow(<Task tasks={tasks}
                                      getTasks={jest.fn()}
                                      updateTask={updateTask}
                                      createTask={jest.fn()}
                                      currentUser={'fake'}
                                />);

        wrapper.find('#tasks-button').simulate('click');
        wrapper.find('#tasks-button').simulate('click');
        expect(wrapper.find('tr').length).toBe(3);
    });
});
