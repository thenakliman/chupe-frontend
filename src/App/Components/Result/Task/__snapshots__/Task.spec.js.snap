// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Task result component snapshot should match the snapshot 1`] = `
<div
  className="task-result-container"
>
  <span
    className="task-create-container"
    id="task-create-container"
  >
    <input
      className="task-input-description"
      id="task-input-description"
      onChange={[Function]}
      value=""
    />
    <button
      className="task-button"
      id="task-create-button"
      onClick={[Function]}
    >
      Create
    </button>
  </span>
  <span>
    <div
      className="all-task-button-container"
    >
      <button
        className="task-button"
        id="tasks-button"
        onClick={[Function]}
      >
        All Tasks
      </button>
    </div>
    <table
      className="team-fund-table-id"
      id="team-fund-table-id"
    >
      <thead>
        <tr>
          <th>
            S.No
          </th>
          <th>
            Description
          </th>
          <th>
            Created
          </th>
          <th>
            In Progress
          </th>
          <th>
            On Hold
          </th>
          <th>
            Done
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          key="0"
        >
          <td>
            0
          </td>
          <td>
            task2
          </td>
          <td
            key="CREATED"
          >
            <input
              checked={false}
              className="task-state-radio-button"
              id="task-state-radio-button-CREATED-undefined"
              name="state-0"
              onChange={[Function]}
              type="radio"
            />
          </td>
          <td
            key="IN_PROGRESS"
          >
            <input
              checked={true}
              className="task-state-radio-button"
              id="task-state-radio-button-IN_PROGRESS-undefined"
              name="state-0"
              onChange={[Function]}
              type="radio"
            />
          </td>
          <td
            key="ON_HOLD"
          >
            <input
              checked={false}
              className="task-state-radio-button"
              id="task-state-radio-button-ON_HOLD-undefined"
              name="state-0"
              onChange={[Function]}
              type="radio"
            />
          </td>
          <td
            key="DONE"
          >
            <input
              checked={false}
              className="task-state-radio-button"
              id="task-state-radio-button-DONE-undefined"
              name="state-0"
              onChange={[Function]}
              type="radio"
            />
          </td>
        </tr>
        <tr
          key="1"
        >
          <td>
            1
          </td>
          <td>
            task1
          </td>
          <td
            key="CREATED"
          >
            <input
              checked={false}
              className="task-state-radio-button"
              id="task-state-radio-button-CREATED-undefined"
              name="state-1"
              onChange={[Function]}
              type="radio"
            />
          </td>
          <td
            key="IN_PROGRESS"
          >
            <input
              checked={true}
              className="task-state-radio-button"
              id="task-state-radio-button-IN_PROGRESS-undefined"
              name="state-1"
              onChange={[Function]}
              type="radio"
            />
          </td>
          <td
            key="ON_HOLD"
          >
            <input
              checked={false}
              className="task-state-radio-button"
              id="task-state-radio-button-ON_HOLD-undefined"
              name="state-1"
              onChange={[Function]}
              type="radio"
            />
          </td>
          <td
            key="DONE"
          >
            <input
              checked={false}
              className="task-state-radio-button"
              id="task-state-radio-button-DONE-undefined"
              name="state-1"
              onChange={[Function]}
              type="radio"
            />
          </td>
        </tr>
      </tbody>
    </table>
  </span>
</div>
`;
